connectors/byoc.py

```diff
@@ -164,22 +167,20 @@ def build_docs_query(self, native_service_types=None, connectors_ids=None):
 
     def _create_object(self, doc_source):
         return Connector(
-            self,
-            doc_source["_id"],
-            doc_source["_source"],
+            doc_source,
             bulk_options=self.bulk_options,
         )
 
 
 class SyncJob:
-    def __init__(self, elastic_index, connector_id, doc_source=None):
```

Comment:
Before it's getting too complicated, I would like to get the team's opinion on this. 

Currently each entity instance (of class `Connector`/`SyncJob`) holds an instance variable `index` (instance of `ConnectorIndex`/`SyncJobIndex`, which is the child class of `ESClient`), so all API calls to Elasticsearch are through this variable).

I want to introduce some static methods, e.g. `pending_jobs` in `SyncJob`, so I can get all pending jobs by calling `SyncJob.pending_jobs()`. But this won't work as the class doesn't hold the index instance. That's why I introduced a class variable (same in class `Connector`), and they should be set in `cli.py`.

cc @elastic/ingestion-team

Reply:
Should `pending_jobs ` be a part of `SyncJobIndex`? `SyncJobIndex` has the ability to query jobs and then create a list of `SyncJob` objects based on the elasticsearch documents (see [_create_object](https://github.com/elastic/connectors-python/blob/main/connectors/byoc.py#L702).

I think `SyncJobIndex` should be responsible for any sync job lookups and `SyncJob` should represent a single document.

Reply:
OK. That's different from my understanding, and the Ruby implementation.

So you are suggesting, the index class (`ConnectorIndex` and `SyncJobIndex`) should be responsible for any business logic related read/write. The entity class (`Connector` and `SyncJob`) only represent a single document, and is responsible for mutating itself, e.g. `job.claim()`, `job.cancel()`? 


Or do you think we could make it a simple data object, and move those mutating methods to the index class as well? This could simply it significantly as it doesn't have to hold an index instance.

Reply:
> So you are suggesting, the index class (ConnectorIndex and SyncJobIndex) should be responsible for any business logic related read/write. 

Yes, logic like pulling documents and create object iterators, pagination. Everything we need to work with Elasticsech indices. 

an abstract code might look like: 

```
for job in sync_job_index.pending_jobs():
  job.claim()
  # execution
  job.finish()
```

> The entity class (Connector and SyncJob) only represent a single document, and is responsible for mutating itself, e.g. job.claim(), job.cancel()?

Yes!

> Or do you think we could make it a simple data object, and move those mutating methods to the index class as well? This could simply it significantly as it doesn't have to hold an index instance.

I don't think so. I think it's pretty good to access the ES client from the entity classes but we can choose another way to propagate ES connection. Now it's an es_index object which is not the ideal way but this can be easily improved because I encapsulated object creation.

Reply:
in your refactorings, make sure you initialize stuff in `__init__` and not add class variables (as opposed to object variables) -- there are very few cases where we want this.
using class variables means you share the same object across all instances, this adds edge issues

Reply:
Thanks for the comments.

Seems the team is not in favor of class variables. I will make the change accordingly.

Reply:
yeah class variables break the class contract because it becomes a namespace and loses its OOP behaviors. It's very helpful in met-programming but in other cases, not. In Python it's completely fine also to use plain functions when a method would be static. It's cleaner